File: server/internal/models/workspace.go
package models

import (
	"time"
)

// Workspace represents a user's workspace in the system
type Workspace struct {
	ID                 int       `json:"id" db:"id,default" validate:"required,min=1"`
	UserID             int       `json:"userId" db:"user_id" validate:"required,min=1"`
	Name               string    `json:"name" db:"name" validate:"required"`
	CreatedAt          time.Time `json:"createdAt" db:"created_at,default"`
	LastOpenedFilePath string    `json:"lastOpenedFilePath" db:"last_opened_file_path"`

	// Integrated settings
	Theme                string `json:"theme" db:"theme" validate:"required,oneof=light dark"`
	AutoSave             bool   `json:"autoSave" db:"auto_save"`
	ShowHiddenFiles      bool   `json:"showHiddenFiles" db:"show_hidden_files"`
	GitEnabled           bool   `json:"gitEnabled" db:"git_enabled"`
	GitURL               string `json:"gitUrl" db:"git_url,ommitempty" validate:"required_if=GitEnabled true"`
	GitUser              string `json:"gitUser" db:"git_user,ommitempty" validate:"required_if=GitEnabled true"`
	GitToken             string `json:"gitToken" db:"git_token,ommitempty,encrypted" validate:"required_if=GitEnabled true"`
	GitAutoCommit        bool   `json:"gitAutoCommit" db:"git_auto_commit"`
	GitCommitMsgTemplate string `json:"gitCommitMsgTemplate" db:"git_commit_msg_template"`
	GitCommitName        string `json:"gitCommitName" db:"git_commit_name"`
	GitCommitEmail       string `json:"gitCommitEmail" db:"git_commit_email" validate:"omitempty,required_if=GitEnabled true,email"`
}

// Validate validates the workspace struct
func (w *Workspace) Validate() error {
	return validate.Struct(w)
}

// ValidateGitSettings validates the git settings if git is enabled
func (w *Workspace) ValidateGitSettings() error {
	return validate.StructExcept(w, "ID", "UserID", "Theme")
}

// SetDefaultSettings sets the default settings for the workspace
func (w *Workspace) SetDefaultSettings() {

	if w.Theme == "" {
		w.Theme = "dark"
	}

	w.AutoSave = w.AutoSave || false
	w.ShowHiddenFiles = w.ShowHiddenFiles || false
	w.GitEnabled = w.GitEnabled || false

	w.GitAutoCommit = w.GitEnabled && (w.GitAutoCommit || false)

	if w.GitCommitMsgTemplate == "" {
		w.GitCommitMsgTemplate = "${action} ${filename}"
	}
}

File: app/src/types/workspace.ts
import { Theme } from './theme';

export interface WorkspaceSettings {
  theme: Theme;
  autoSave: boolean;
  showHiddenFiles: boolean;
  gitEnabled: boolean;
  gitUrl: string;
  gitUser: string;
  gitToken: string;
  gitAutoCommit: boolean;
  gitCommitMsgTemplate: string;
  gitCommitName: string;
  gitCommitEmail: string;
}

export const DEFAULT_WORKSPACE_SETTINGS: WorkspaceSettings = {
  theme: Theme.Light,
  autoSave: false,
  showHiddenFiles: false,
  gitEnabled: false,
  gitUrl: '',
  gitUser: '',
  gitToken: '',
  gitAutoCommit: false,
  gitCommitMsgTemplate: '${action} ${filename}',
  gitCommitName: '',
  gitCommitEmail: '',
};

export interface Workspace extends WorkspaceSettings {
  name: string;
  createdAt: number;
}

export const DEFAULT_WORKSPACE: Workspace = {
  name: '',
  createdAt: Date.now(),
  ...DEFAULT_WORKSPACE_SETTINGS,
};

export function isWorkspace(obj: unknown): obj is Workspace {
  return (
    typeof obj === 'object' &&
    obj !== null &&
    'name' in obj &&
    typeof (obj as Workspace).name === 'string' &&
    'theme' in obj &&
    typeof (obj as Workspace).theme === 'string' &&
    'autoSave' in obj &&
    typeof (obj as Workspace).autoSave === 'boolean' &&
    'gitEnabled' in obj &&
    typeof (obj as Workspace).gitEnabled === 'boolean' &&
    'gitUrl' in obj &&
    typeof (obj as Workspace).gitUrl === 'string' &&
    'gitUser' in obj &&
    typeof (obj as Workspace).gitUser === 'string' &&
    'gitToken' in obj &&
    typeof (obj as Workspace).gitToken === 'string' &&
    'gitAutoCommit' in obj &&
    typeof (obj as Workspace).gitAutoCommit === 'boolean' &&
    'gitCommitMsgTemplate' in obj &&
    typeof (obj as Workspace).gitCommitMsgTemplate === 'string'
  );
}
